src/main/java/ro/utcn/sd/mdantonio/StackUnderflow/service/PostManagementService.java

```diff
@@ -0,0 +1,141 @@
+@RequiredArgsConstructor
+public class PostManagementService implements ManagementService {
+    private final RepositoryFactory repositoryFactory;
+    @Transactional
+    public List<Post> listQuestions(){
+        List<Post> questions =  repositoryFactory.createPostRepository().findAll().stream().
+                filter(x->x.getPosttypeid()==QUESTIONID).collect(Collectors.toList());
+        questions.forEach(x->x.setScore(calculatePostScore(x)));
+        return questions.stream().sorted(Comparator.comparing(Post::getCreationDate,
+                Comparator.nullsLast(Comparator.reverseOrder()))).
+                sorted((x, y)->Long.compare(y.getScore(), x.getScore())).collect(Collectors.toList());
+    }
+
+    @Transactional
+    public List<Post> listQuestionResponses(int questionId) throws ObjectNotFoundExpection{
+        PostRepository postRepository = repositoryFactory.createPostRepository();
+        postRepository.findById(questionId).orElseThrow(ObjectNotFoundExpection::new);
+        List<Post> answers = postRepository.findAll().stream().filter(x->x.getParentid()!=null).
+                filter(x->x.getParentid().equals(questionId)).collect(Collectors.toList());
+        answers.forEach(x->x.setScore(calculatePostScore(x)));
+        return answers.stream().sorted(Comparator.comparing(Post::getCreationDate,
+                Comparator.nullsLast(Comparator.reverseOrder()))).
+                sorted((x,y)->Long.compare(y.getScore(), x.getScore())).collect(Collectors.toList());
+    }
+
+    @Transactional
+    public Post addPost(int postTypeId, int authorId, Integer parentId, String title, String body, Date creationDate)
+        throws ObjectNotFoundExpection{
+        PostRepository postRepository = repositoryFactory.createPostRepository();
+
+        if(postTypeId==ANSWERID)
+            postRepository.findById(parentId).orElseThrow(ObjectNotFoundExpection::new);
+        repositoryFactory.createUnderflowUserRepository().findById(authorId).orElseThrow(ObjectNotFoundExpection::new);
+
+        return postRepository.save(new Post(postTypeId, authorId, parentId, title, body, creationDate));
+    }
+
+    @Transactional
+    public Vote votePost(int currentUserId, int postId, boolean isUpvote)
+            throws ObjectNotFoundExpection, InvalidActionException{
+        Post post = repositoryFactory.createPostRepository().findById(postId).orElseThrow(ObjectNotFoundExpection::new);
+        UnderflowUser user = repositoryFactory.createUnderflowUserRepository().findById(currentUserId).
+                orElseThrow(ObjectNotFoundExpection::new);
+
+
+        if(user.getUserid().equals(post.getAuthorid()))
+            throw new InvalidActionException();
+
+        return repositoryFactory.createVoteRepository().save(new Vote(currentUserId, postId, isUpvote));
+    }
+
+    @Transactional
+    public List<Post> listPostByTitle(String title){
+        List<Post> postList =  repositoryFactory.createPostRepository().findAll().stream()
```

Comment:
It would have been better if you had the filtering at the level of the repositories. Imagine that you have a large database -> you would potentially load millions of posts just to find a few specific ones.

Reply:
This was taken into consideration when I started doing this :))) . I really hope I won't have to make 4 implementations for the same type of database in 2-3 days in the real world =)).

Reply:
:) Yes, it is true that in real life you would not have to implement 4 different types of repositories. Nevertheless, it would be a good exercise to try to move this filtering logic down to the repository level.